pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+// syncConsistentHashByEIP refresh the consistentHashMap
+func (c *Cluster) syncConsistentHashByEIP(eip *v1alpha2.ExternalIPPool) error {
+	nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+	if err != nil {
+		return err
+	}
+	nodes, err := c.nodeLister.List(nodeSel)
+	if err != nil {
+		return err
+	}
+	c.consistentHashMap[eip.Name] = newNodeConsistentHashMap()
+	aliveNodes := c.aliveNodeList()
+	// node alive and node labels matches with externalIPPool nodeSelector
+	aliveAndMatchedNodes := sets.NewString()
+	for _, node := range nodes {
+		nodeName := node.Name
+		if aliveNodes.Has(nodeName) {
+			aliveAndMatchedNodes.Insert(nodeName)
+		}
+	}
+	c.consistentHashMap[eip.Name].Add(aliveAndMatchedNodes.List()...)
+	c.notify(eip.Name)
+	return nil
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Handle clusterNode events failed")
+	}
+	switch node, event := nodeEvent.Node, nodeEvent.Event; event {
+	case memberlist.NodeJoin:
+		// when node join cluster, all matched externalIPPools consistentHash should changed
+		node, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			klog.ErrorS(err, "Processing clusterNode events failed")
+			return
+		}
+		count := 0
+		for _, eip := range eips {
+			if c.matchNodeSelector(node, eip) {
+				c.queue.Add(eip.Name)
+				klog.V(5).InfoS("Node join event", "nodeName", node.String(), "eip", eip.Name)
+				count++
+			}
+		}
+		klog.InfoS("Node event: join node", "nodeName", node.Name, "eipNum", count)
+	case memberlist.NodeLeave:
+		// when node leave cluster, the node may failed or deleted, all externalIPPools consistentHash maybe changed
+		for _, eip := range eips {
+			c.queue.Add(eip.Name)
+			klog.V(5).InfoS("Node leave event", "nodeName", node.String(), "eip", eip.Name)
+		}
+		klog.InfoS("Node event: leave node", "nodeName", node.Name, "eipNum", len(eips))
+	case memberlist.NodeUpdate:
+		klog.InfoS("Node event: update node", "nodeName", node.Name)
+	default:
+		klog.InfoS("Node event: unknown", "nodeName", node.Name, "event", event)
+	}
+}
+
+// aliveNodeList return alive nodeName list in cluster
+func (c *Cluster) aliveNodeList() sets.String {
+	aliveMembers := c.mList.Members()
+	nodes := make([]string, len(aliveMembers))
+	for i, node := range aliveMembers {
+		nodes[i] = node.Name
+	}
+	return sets.NewString(nodes...)
+}
+
+func (c *Cluster) shouldSelect(consistentHash *consistenthash.Map, name string) bool {
+	myNode := c.nodeName
+	hitted := hitNodeByConsistentHash(consistentHash, name, myNode)
+	klog.InfoS("Assign egress owner for local node", "egressName", name, "localNode", myNode, "shouldSelected", hitted)
```

Comment:
I believe this entry will be repeated in logs whenever a pod applies the egress, which is very duplicate. I think we should let callers to decide whether to log it and it should log it only when there is a change.

Reply:
done

Reply:
I don't mean the log level, I mean it should logged in `(c *EgressController) syncEgress` when it detects a change like below, because `shouldSelect` and `ShouldSelectEgress` is a readonly method and shouldn't log anything it doesn't do, i.e. assigning IP is not in its scope`, and the method can be called to check whether the egress's owner changes, so definitely misleading if it's logged everytime when there's no change.
```
if c.cluster.ShouldSelectEgress(egress) {
	if !c.localIPDetector.IsLocalIP(egress.Spec.EgressIP) {
		klog.InfoS("Assign Egress IP to this Node", ...)
		err := c.ipAssigner.AssignEgressIP(egress.Spec.EgressIP)
		if err != nil {
			return err
		}
	}
	if err := c.updateEgressStatus(egress, c.nodeName); err != nil {
		return fmt.Errorf("update egress status error:%v", err)
	}
```

Reply:
done